src/python_testing/TC_CC_2_1.py

```diff
@@ -0,0 +1,419 @@
+                asserts.fail("Range list is empty")
+        elif data_type == self.STRING and isinstance(attr_val, str):
+            if max_len > 0:
+                logger.info(f"Validating string with a max len of {max_len}")
+                asserts.assert_true((len(attr_val) <= max_len), "String len is out of range.")
+            else:
+                asserts.fail("FAILED")
+        else:
+            asserts.fail("Validation not possible as not data type provided.")
+
+        # if we land at this point it mean validation had passed
+        # check if string has uint and verify is we need to compare against a min or max value
+        if 'uint' in data_type and (max_len is not None or min_len is not None):
+            if isinstance(max_len, int):
+                logger.info(f"Max len defined validation max range for uint {max_len}")
+                asserts.assert_true((attr_val <= max_len), f"Attribute {attribute} is out of range (max): {max_len}")
+            if isinstance(min_len, int):
+                logger.info(f"Min len defined validation max range for uint {max_len}")
+                asserts.assert_true((attr_val >= min_len), f"Attribute {attribute} is out of range (min): {min_len}")
+
+        return attr_val
+
+    def _verify_for_numberofprimaries_value(self, numberofprimaries_value: Optional[int] = None, numberofprimaries_condition: int = 0) -> bool:
+        """Verify the numbrofprimaries attribute value against the condition. Return the status to skip tests.
+        Args:
+            numberofprimaries_value (Optional[int], optional): _description_. Defaults to None.
+            numberofprimaries_condition (int, optional): _description_. Defaults to 0.
+        Returns:
+            bool: Return the status of skipping verification.
+        """
+        skip_tests = False
+        logger.info(
+            f"Verifying the value of numberofprimaries {numberofprimaries_value} is greater than {numberofprimaries_condition} ?")
+        # is not equal or greater than
+        if numberofprimaries_value < numberofprimaries_condition:
+            skip_tests = True
+
+        return skip_tests
+
+    def _verify_first_4bits(self, numa, numb):
+        # get the first 4 bits and compare them
+        tmp_a = numa & (2**4-1)
+        tmp_b = numb & (2**4-1)
+        logger.info("Verifying if first 4 bits are equal.")
+        logger.info(f"Num a : {bin(tmp_a)}")
+        logger.info(f"Num b : {bin(tmp_b)}")
+        asserts.assert_equal(tmp_a, tmp_b, "Numbers are not equal in bit after bit comparsion")
+
+    @async_test_body
+    async def test_TC_CC_2_1(self):
+        self.UINT8 = "uint8"
+        self.UINT16 = "uint16"
+        self.UINT32 = 'uint32'
+        self.STRING = "string"
+        self.ENUM = "enum"
+        self.cluster = Clusters.ColorControl
+        self.endpoint = self.get_endpoint(1)
+
+        self.step(1)
+
+        self.step(2)
+        await self._verify_attribute("CurrentHue", self.UINT8)
```

Comment:
Would you mind opening a follow up issue to convert this test over to use the data from the data model XML files to get the constraints? This definitely matches the test plan, and we should get this in, but seeing all these hard coded values seems like a recipe for bugs.